Сдача работы на проверку

sergiopetrovykh opened the pull request.

sergiopetrovykh committed:
1_commit

Написал код по 1 части задания

sergiopetrovykh committed:
2_commit

Обновлен package-lock.json

sergiopetrovykh committed:
3_commit

обновил webpack.config
обновил слайсы. возникла ошибка Can't resolve './slices/orderSlice'

sergiopetrovykh committed:
4

Убрал критическую ошибку

sergiopetrovykh committed:
5

sergiopetrovykh committed:
6

sergiopetrovykh committed:
7

объединил basketReducer и constructorSlice в один файл, как нужно было. Посыпались ошибки

sergiopetrovykh committed:
7

sergiopetrovykh committed:
7

sergiopetrovykh committed:
8

sergiopetrovykh committed:
8.1

sergiopetrovykh committed:
9

Работает конструктор, лента заказов, профиль, вход в профиль.
НЕ работает "оформить заказ"

sergiopetrovykh committed:
10

Починил кнопку "Оформить заказ"

aromanovnik commented on src/components/burger-ingredient/burger-ingredient.tsx:
```diff
@@ -3,12 +3,17 @@ import { useLocation } from 'react-router-dom';
+import { useDispatch } from 'react-redux';
 
 export const BurgerIngredient: FC<TBurgerIngredientProps> = memo(
   ({ ingredient, count }) => {
```
**Надо исправить**: У добавленных в бургер ингредиентов должен отображаться счётчик с количеством добавленных в бургер ингредиентов этого типа. При добавлении ингредиента в бургер значение счётчика увеличивается, при удалении — значение счётчика уменьшается.

<img width="1330" alt="Screenshot 2024-11-20 at 15 30 04" src="https://github.com/user-attachments/assets/7ab17f6f-47c3-4d13-b855-74033d83f0dd">

aromanovnik commented on src/components/protected-route/protected-route.tsx:
```diff
@@ -0,0 +1,27 @@
+    return <Navigate to='/login' state={{ from: location }} />;
+  }
+
+  return <>{children}</>;
```
**Надо исправить**: По чек-листу разметка компонентов должна быть заключена в круглые скобки `()`, чтобы улучшить читаемость и избежать ошибок синтаксиса при многосрочных JSX-элементах. Это является хорошей практикой в React, особенно когда компонент содержит несколько строк или сложную разметку.

aromanovnik commented on src/components/protected-route/protected-route.tsx:
```diff
@@ -0,0 +1,27 @@
+import React, { FC, useEffect } from 'react';
+import { Navigate, useLocation } from 'react-router-dom';
+import { useSelector, useDispatch } from 'react-redux';
+// import { checkAuthStatus } from '../../services/slices/user';
```
**Можно лучше**: Закомментированный код лучше удалять, так как он засоряет проект, снижает его читаемость и может вводить в заблуждение. История изменений сохраняется в системе контроля версий, такой как Git, где всегда можно восстановить удаленный код, если он потребуется.

aromanovnik commented on src/components/protected-route/protected-route.tsx:
```diff
@@ -0,0 +1,27 @@
+  children,
+  onlyUnAuth
+}) => {
+  const isAuthorized = useSelector((state: any) => state.user.isAuthorized);
```
**Надо исправить**: Нужно использовать селектор `useSelector(isAuthorizedSelector)`, а не `useSelector((state) => state.user.isAuthorized)`. Глобальный стейт лучше не использовать напрямую в компонентах, так как это может привести к дублированию логики. Вместо этого рекомендуется использовать созданные селекторы. Это помогает улучшить читаемость и поддержку кода, так как селекторы инкапсулируют логику доступа к данным, обеспечивают переиспользуемость и упрощают тестирование.

aromanovnik commented on src/components/protected-route/protected-route.tsx:
```diff
@@ -0,0 +1,27 @@
+  }
+
+  if (!isAuthorized && !onlyUnAuth) {
+    return <Navigate to='/login' state={{ from: location }} />;
```
**Надо исправить**: При заходе на страницу `/profile/orders` авторизованным пользователем сначала происходит редирект на `/login`, а потом снова на `/profile/orders`. Вместо этого, пока информация о пользователе не загружена, должен отображаться прелоадер.

aromanovnik commented on src/services/slices/orderSlice.ts:
```diff
@@ -0,0 +1,49 @@
+};
+
+// Асинхронный thunk для получения заказа по номеру
+export const getOrderThunk = createAsyncThunk(
```
**Можно лучше:** Если в `createAsyncThunk` нужно просто вызвать функцию запроса и получить возвращаемое знание, без какой либо его обработки, то не нужно создавать функцию обертку.
Лучше просто передать саму функцию запроса в качестве аргумента, например: `createAsyncThunk('feed/getOrder', getOrderByNumberApi)`

aromanovnik commented on src/components/app/app.tsx:
```diff
@@ -1,14 +1,149 @@
+            path='/feed/:number'
+            element={
+              <Modal
+                title={`#${useParams().number}`} // Использование useParams
```
**Можно лучше**: Чтобы избежать значения `undefined`, лучше использовать `useMatch`, а не `useParams()` в главном компоненте App, так как `useParams()` возвращает параметры только для активного маршрута, а в сложной вложенной структуре маршрутов это может привести к ошибкам. `useMatch` же позволяет надежно проверять совпадение текущего маршрута с заданным шаблоном, даже если маршрут вложен, что делает проверку параметров более предсказуемой и удобной. Например для `feed`: `const feedMatch = useMatch('/feed/:number')?.params.number;`

<img width="907" alt="Screenshot 2024-11-20 at 17 24 53" src="https://github.com/user-attachments/assets/1f0916b0-1e40-46b9-8433-41a832f1027c">

aromanovnik reviewed (changes requested):
Добрый день, Сергей!

Работа выполнена очень хорошо: отображается имя пользователя в шапке, роуты `/login`, `/register`, `/forgot-password` и `/reset-password` недоступны авторизованным пользователям, в шапке сайта корректно отображается активной та ссылка, где сейчас находится пользователь, конструктор бургера работает правильно, заказ оформляется, модальные окна с информацией о заказах открываются по отдельному пути. Хорошее понимание Redux. Код выглядит чисто и аккуратно.

Тем не менее, есть несколько замечаний, которые потребуют доработки. Более подробно о них написано в комментариях к коду.

Что нужно исправить:
* Нужно использовать селектор `useSelector(isAuthorizedSelector)`, а не `useSelector((state) => state.user.isAuthorized)`. Глобальный стейт лучше не использовать напрямую в компонентах.
* У добавленных в бургер ингредиентов должен отображаться счётчик с количеством добавленных в бургер ингредиентов этого типа.
* При заходе на страницу `/profile/orders` авторизованным пользователем сначала происходит редирект на `/login`, а потом снова на `/profile/orders`. Вместо этого, пока информация о пользователе не загружена, должен отображаться прелоадер. 
* По чек-листу разметка компонентов должна быть заключена в круглые скобки `()`.

Что можно улучшить:
* Чтобы избежать значения `undefined`, лучше использовать `useMatch`, а не `useParams()` в главном компоненте App, так как `useParams()` возвращает параметры только для активного маршрута, а в сложной вложенной структуре маршрутов это может привести к ошибкам. 
* Закомментированный код лучше удалять, так как он засоряет проект, снижает его читаемость и может вводить в заблуждение.
* Если в `createAsyncThunk` нужно просто вызвать функцию запроса и получить возвращаемое знание, без какой либо его обработки, то не нужно создавать функцию обертку.

Напоминаю, что работа может быть принята только после исправления всех критических замечаний Надо исправить.

Удачного рефакторинга кода.

sergiopetrovykh committed:
11

правки после ревью
-селектор useSelector(isAuthorizedSelector), а не useSelector((state) => state.user.isAuthorized). Глобальный стейт лучше не использовать напрямую в компонентах.
-У добавленных в бургер ингредиентов настроено отображение счётчика с количеством добавленных в бургер ингредиентов этого типа.
- При заходе на страницу /profile/orders авторизованным пользователем сначала происходит редирект на /login, а потом снова на /profile/orders. Вместо этого, пока информация о пользователе не загружена, должен отображаться прелоадер.
- По чек-листу разметка компонентов заключена в круглые скобки ().

sergiopetrovykh committed:
11.1

aromanovnik reviewed (approved):
Поздравляю! Ваша работа принята. Все недочеты исправлены. 

**Можно лучше**: сейчас `lint` выдает предупреждения. Хорошей практикой считается запускать `lint` и `prettier` перед коммитом. Для удобства эти команды можно объединить в одну, например: `"pre-commit": "npm run format && npm run lint:fix"`. https://disk.yandex.ru/i/4jNv0A_ydGhvgA

Желаю успехов в дальнейшей учебе!

sergiopetrovykh committed:
11.2

Исправлено замечание ревьюера:
Можно лучше: сейчас lint выдает предупреждения. Хорошей практикой считается запускать lint и prettier перед коммитом. Для удобства эти команды можно объединить в одну, например: "pre-commit": "npm run format && npm run lint:fix".

sergiopetrovykh committed:
11-3

sergiopetrovykh committed:
11-4

sergiopetrovykh committed:
11-5

sergiopetrovykh committed:
11-5

sergiopetrovykh committed:
11-6

sergiopetrovykh committed:
11-6
